posts/avoid-timeouts-azure-data-factory.md

```diff
@@ -0,0 +1,201 @@
+
+6.  Enter a SQL query to select the range of data using the limit and offset variables created in step 5. 
+
+Note, this is for a MySQL database and the SQL query will likely change for a different source SQL database.
+
+Here's what I'm using in this example:
+
+```
+select 
+    *
+from adf_main.users
+where id > @{variables('offset_start')}
+limit @{variables('limit')}
+```
+
+![model_6](https://github.com/portillosc/blog-usa/blob/master/images/2022/01/adf-timeouts-6.png)
+
+<p align = "center">
+Figure 6: Adding a recurring "Copy data" activity
+</p>
+<br/>
+
+1.  Create a "Lookup" activity.
+
+2.  Select the same source as before as well as the "Query" option.
+
+3.  Add a SQL statement that retrives the last ID from the set of results.
+
+Here's what I'm using in Figure 7:
+
+```
+select
+    max(t.id) as last_id
+from (
+    select
+        id
+    from adf_main.users
+    where id > @{variables('offset_start')}
+    limit @{variables('limit')}
+) as t;
+```
+
+![model_7](https://github.com/portillosc/blog-usa/blob/master/images/2022/01/adf-timeouts-7.png)
+
+<p align = "center">
+Figure 7: A lookup activity to retrieve the last ID from the results
+</p>
+<br/>
+
+1.  Add a "Set variable" activity to update the "offset_start" variable to the last seen ID. This allows the next set of results to pick up from where the previous left off.
+
+Here's the value used in Figure 8:
+
+```
+@string(activity('users last_id').output.firstRow.last_id)
+```
+
+This is taking the output from the "Lookup" activity and setting it directly to the "offset_start" variable.
+
+![model_8](https://github.com/portillosc/blog-usa/blob/master/images/2022/01/adf-timeouts-8.png)
+
+<p align = "center">
+Figure 8: "Set variable" activity to update the offset_start variable
+</p>
+<br/>
+
+1.  Move back out into the "Until" function and set the end condition.
+
+This is the value in Figure 9:
+
+```
+@less(activity('copy users range').output.rowsCopied, 100)
+```
+
+![model_9](https://github.com/portillosc/blog-usa/blob/master/images/2022/01/adf-timeouts-9.png)
+
+<p align = "center">
+Figure 9: Setting the end condition for the "Until" function
+</p>
+<br/>
+
+This end condition tests to see whether the output of the inner "Copy data" activity is less than the limit that was set in the beginning.
+
+If it is not, then that means there is more data to copy and the offset will increment.
+
+Otherwise, the "Until" function will end.  
+<br/>
+
+## Results and performance
+This is the result in Azure Storage of copying the adf_main.users table by 100 records at a time.
+
+![model_10](https://github.com/portillosc/blog-usa/blob/master/images/2022/01/adf-timeouts-10.png)
+
+<p align = "center">
+Figure 10: Azure Storage output from the ADF iterative range copy
+</p>
+<br/>
+
+The Parquet file format was chosen since its a good format for Data Lake storage and easily consumed for downstream transformations by Spark and Data Factory Data Flows. 
```

Comment:
```suggestion
The Parquet file format was chosen since it is a good format for Data Lake storage and easily consumed for downstream transformations by Spark and Data Factory data flows. 
```